lib/folio_client/data_import.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require "timeout"
+require "date"
+require "dry/monads"
+require "marc"
+require "stringio"
+
+class FolioClient
+  # Imports MARC records into FOLIO
+  class DataImport
+    include Dry::Monads[:result]
+
+    attr_reader :job_execution_id
+
+    # @param client [FolioClient] the configured client
+    # @param record [MARC::Record] record to be imported
+    # @param job_profile_id [String] job profile id to use for import
+    # @param job_profile_name [String] job profile name to use for import
+    def initialize(client, marc:, job_profile_id:, job_profile_name:)
+      @client = client
+      @marc = marc
+      @job_profile_id = job_profile_id
+      @job_profile_name = job_profile_name
+    end
+
+    def import
+      response_hash = client.post("/data-import/uploadDefinitions", {fileDefinitions: [{name: marc_filename}]})
+      upload_definition_id = response_hash.dig("fileDefinitions", 0, "uploadDefinitionId")
+      @job_execution_id = response_hash.dig("fileDefinitions", 0, "jobExecutionId")
+      file_definition_id = response_hash.dig("fileDefinitions", 0, "id")
+
+      upload_file_response_hash = client.post("/data-import/uploadDefinitions/#{upload_definition_id}/files/#{file_definition_id}", marc_binary(marc), content_type: "application/octet-stream")
+
+      client.post("/data-import/uploadDefinitions/#{upload_definition_id}/processFiles",
+        {uploadDefinition: upload_file_response_hash, jobProfileInfo: {id: job_profile_id, name: job_profile_name, dataType: "MARC"}})
+    end
+
+    # @return [Dry::Monads::Result] Success if job is complete, Failure(:pending) if job is still running, Failure(:error) if job has errors
+    def job_status
+      response_hash = client.get("/metadata-provider/jobSummary/#{job_execution_id}")
+
+      return Failure(:error) if response_hash["totalErrors"].positive?
+      return Failure(:pending) if response_hash.dig("sourceRecordSummary", "totalCreatedEntities").zero? && response_hash.dig("sourceRecordSummary", "totalUpdatedEntities").zero?
```

Comment:
When I'm working with the API, I get a 404 initially when the processing is just getting started.  The response body is `JobSummary for jobExecutionId: 'bae2f377-6627-40af-94b8-a3a957baf055' was not found`. If there is not a JSON body, what will the result be here? I didn't realize this when writing up the ticket, sorry.